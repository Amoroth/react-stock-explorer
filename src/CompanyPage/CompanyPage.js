import React, { Component } from 'react'
import PropTypes from 'prop-types'

import PageTitle from '../shared/PageTitle'
import styles from './CompanyPage.module.css'

class CompanyPage extends Component {
  state = {
    company: {
      companyName: '',
      symbol: '',
      exchange: '',
      industry: '',
      website: '',
      description: '',
      CEO: '',
    },
    logo: '',
    financials: [
      {
        reportDate: '',
        grossProfit: 0,
        totalRevenue: 0,
        netIncome: 0,
        researchAndDevelopment: 0,
        totalAssets: 0,
        totalDebt: 0,
      },
    ],
    finTime: 0,
  }

  componentDidMount() {
    const { location } = this.props
    const companyQuery = new URLSearchParams(location.search.slice(1)).get(
      'cmp',
    )
    fetch(
      `https://api.iextrading.com/1.0/stock/${companyQuery}/batch?types=company,logo,financials`,
    )
      .then((res) => res.json())
      .then((json) => {
        this.setState({
          company: json.company,
          logo: json.logo.url,
          financials: json.financials.financials,
        })
      })
  }

  onSelectFintime = (index) => {
    this.setState({ finTime: index })
  }

  render() {
    const { financials, finTime, logo, company } = this.state
    const { favorites, onFavorite, currency, currencyFormat } = this.props

    let currentReport = {}
    if (financials) {
      currentReport = Object.assign({}, financials[finTime])
      delete currentReport.reportDate
    }

    let financialsElement = null
    if (financials) {
      financialsElement = (
        <React.Fragment>
          <hr />
          <div>
            <div className={styles['fintime-container']}>
              {financials.map((val, ind) => (
                <button
                  className={`${styles['fintime-button']} ${
                    finTime === ind ? styles['fintime-button-active'] : null
                  }`}
                  type="button"
                  onClick={() => this.onSelectFintime(ind)}
                  key={val.reportDate}
                >
                  {val.reportDate}
                </button>
              ))}
            </div>
            <div className={styles['fin-info']}>
              {Object.keys(currentReport).map((val) => {
                const tempName = val.replace(/([A-Z])/g, ' $1')
                const labelName = tempName.charAt(0).toUpperCase() + tempName.slice(1)
                return (
                  <React.Fragment key={val}>
                    <span>{labelName}</span>
                    <span>{currentReport[val] ? `${currencyFormat(currentReport[val])} ${currency}` : '-'}</span>
                  </React.Fragment>
                )
              })}
            </div>
          </div>
        </React.Fragment>
      )
    }

    return (
      <div className={styles.container}>
        <PageTitle
          logo={logo}
          name={company.companyName}
          exchange={company.exchange}
          symbol={company.symbol}
          favorite={favorites.includes(company.symbol)}
          onFav={onFavorite}
        />
        <hr />
        <div>
          <div className={styles['company-info-row']}>
            <span>
              {`CEO: ${company.CEO}`}
            </span>
            <span>
              {`Industry: ${company.industry}`}
            </span>
            <span>
              <a href={company.website}>Website</a>
            </span>
          </div>
          <p>{company.description || 'No Description.'}</p>
        </div>
        {financialsElement}
      </div>
    )
  }
}

CompanyPage.propTypes = {
  location: PropTypes.objectOf(PropTypes.string).isRequired,
  favorites: PropTypes.arrayOf(PropTypes.string).isRequired,
  onFavorite: PropTypes.func.isRequired,
  currency: PropTypes.string.isRequired,
  currencyFormat: PropTypes.func.isRequired,
}

export default CompanyPage
