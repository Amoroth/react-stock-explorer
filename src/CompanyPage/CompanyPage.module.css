.container {
  margin: auto;
  width: 60%;
  max-width: 800px;
  background-color: #fff;
  border-radius: 5px;
  box-shadow: 2px 4px 10px rgba(0,0,0,.025);
  padding: 16px;
}

.company-info-row {
  display: flex;
  justify-content: space-between;
}

.container a {
  color: #344955;
  transition: color .2s;
}

.container a:hover {
  color: #75b8e2;
}

@media screen and (max-width: 620px) {
  .container {
    width: 97%;
  }

  .company-info-row {
    flex-direction: column;
  }

  .company-info-row span {
    margin: 10px 0;
  }
}