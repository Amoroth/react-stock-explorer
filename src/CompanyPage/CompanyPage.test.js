import React from 'react'
import { shallow } from 'enzyme'

import CompanyPage from './CompanyPage'

describe('<CompanyPage />', () => {
  let props
  let wrapper

  beforeEach(() => {
    props = {
      location: { search: '&cmp=COM' },
      favorites: [],
      onFavorite: jest.fn(),
      currency: 'USD',
      currencyFormat: jest.fn(),
    }
    wrapper = shallow(<CompanyPage {...props} />)
  })

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('fetches on mount', () => {
    expect(fetch).toBeCalledWith('https://api.iextrading.com/1.0/stock/COM/batch?types=company,logo,financials')
  })

  it('renders page title', () => {
    expect(wrapper.find('withRouter(pageTitle)').exists()).toBe(true)
  })

  it('renders CEO, Industry and website', () => {
    wrapper.setState({
      company: {
        CEO: 'Guy',
        industry: 'Space',
        website: 'ehh',
      },
    })

    const spans = wrapper.find('.company-info-row').children()
    expect(spans.length).toBe(3)
    expect(spans.at(0).text()).toBe('CEO: Guy')
    expect(spans.at(1).text()).toBe('Industry: Space')
    expect(spans.at(2).children().first().type()).toBe('a')
    expect(spans.at(2).children().first().prop('href')).toBe('ehh')
  })

  it('chenges time of finances', () => {
    wrapper.setState({ financials: [{ reportDate: '08.12.2018' }, { reportDate: '09.12.2018' }] })

    wrapper.find('.fintime-button').at(1).simulate('click')
    expect(wrapper.state('finTime')).toBe(1)
  })

  it('chegest fintime button class on active', () => {
    wrapper.setState({ financials: [{ reportDate: '08.12.2018' }, { reportDate: '09.12.2018' }] })

    expect(wrapper.find('.fintime-button').at(0).prop('className')).toBe('fintime-button fintime-button-active')
    expect(wrapper.find('.fintime-button').at(1).prop('className')).toBe('fintime-button null')
    wrapper.find('.fintime-button').at(1).simulate('click')
    expect(wrapper.find('.fintime-button').at(0).prop('className')).toBe('fintime-button null')
    expect(wrapper.find('.fintime-button').at(1).prop('className')).toBe('fintime-button fintime-button-active')
  })

  it('renders financial details', () => {
    wrapper.setState({ financials: [{ reportDate: '08.12.2018', a: 1, b: 2, c: null }] })

    expect(wrapper.find('.fin-info').children().length).toBe(6) // No report date
    expect(wrapper.find('.fin-info').children().last().text()).toBe('-')
    expect(props.currencyFormat.mock.calls.length).toBe(2) // not called on empty values
  })
})
