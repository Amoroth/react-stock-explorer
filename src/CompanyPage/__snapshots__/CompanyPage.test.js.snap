// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CompanyPage /> renders 1`] = `
<div
  className="container"
>
  <withRouter(pageTitle)
    exchange=""
    favorite={false}
    logo=""
    name=""
    onFav={[MockFunction]}
    symbol=""
  />
  <hr />
  <div>
    <div
      className="company-info-row"
    >
      <span>
        CEO: 
      </span>
      <span>
        Industry: 
      </span>
      <span>
        <a
          href=""
        >
          Website
        </a>
      </span>
    </div>
    <p>
      No Description.
    </p>
  </div>
  <hr />
  <div>
    <div
      className="fintime-container"
    >
      <button
        className="fintime-button fintime-button-active"
        key=""
        onClick={[Function]}
        type="button"
      />
    </div>
    <div
      className="fin-info"
    >
      <span>
        Gross Profit
      </span>
      <span>
        -
      </span>
      <span>
        Total Revenue
      </span>
      <span>
        -
      </span>
      <span>
        Net Income
      </span>
      <span>
        -
      </span>
      <span>
        Research And Development
      </span>
      <span>
        -
      </span>
      <span>
        Total Assets
      </span>
      <span>
        -
      </span>
      <span>
        Total Debt
      </span>
      <span>
        -
      </span>
    </div>
  </div>
</div>
`;
